### PR TITLE
networkd: force bringing up devices with no IP addresses

### PR DESCRIPTION
**Pull request by cyphermox**
Signed-off-by: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>

### CODE REVIEW

**Comment by cyphermox** [comment]
This is failing tests on purpose for now, as it's up for review but might need more work to make sure things are working correctly.

The intent is to fix bringing up anonymous bridges, it seems to work reasonably well for that. Non-virtual interfaces will also be affected though, and we should make sure they are fine to be UP, NO-CARRIER in this case. Where people absolutely want the devices to be unconfigured, then we should expect the devices to not be listed at all in netplan YAML.

File: src/networkd.c [hunk 1]
```diff
@@ -291,69 +291,71 @@ write_ip_rule(ip_rule* r, GString* s)
...
+    /* Prepare the [Network] section */
+    network = g_string_sized_new(200);
 
-    if (def->optional)
-        g_string_append(s, "\n[Link]\nRequiredForOnline=no\n");
+    /* Enforce bringing up devices with no configured addresses. This is
+       especially useful for bringing up anonymous bridges
+     */
+    if (!def->dhcp4 && !def->dhcp6 && !def->ip4_addresses && !def->ip6_addresses)
+	g_string_append(network, "ConfigureWithoutCarrier=true\n");
```
[the remark this answers is not in the document]
Comment (cyphermox) [reply src/networkd.c]: That's a different issue; a bridge with no members will be handled separately in a rework of the validation code.

Here, it's exactly what we want to "configure" the interface is there's no carrier and no address; and afterwards we'll add "configuring with no carrier" for the interface-less case.

**Comment by cyphermox** [comment]
Merged as of https://github.com/CanonicalLtd/netplan/commit/a5389e579a73d2b45d1f5040dedad39580285602

**Comment by KalenAnson** [comment]
@cyphermox Looking this closed PR over, I see that e681f1ea075625a7de87a966f8f460a387fc3266 did not make the cut (or was unintentionally omitted). I am referring specifically to configuring a physical device without a carrier. Master Ref: https://github.com/CanonicalLtd/netplan/blob/master/src/networkd.c#L576

The update to the documentation seems to indicate that the change in behavior in this PR would include the condition that the linked commit addresses, i.e. an interface specified in the netplan config will be configured by the renderer (regardless of carrier link state presumably).

> If an interface is defined with an ID in a configuration file; it will be brought up by the applicable renderer.  To not have netplan touch an interface at all, it should be  completely omitted from the netplan configuration files.

Can you shed some light on why this additional behavior change was not introduced to how netplan functions?

It would be very convenient for an interface that has no carrier to still be able to receive a static address so that services that depend on it would not have to specify interface requirements in both their own config files as well as systemd service files in the form of 'after's, 'bindsto's and 'wants' (dhcp servers, dns servers, et.al.).

**Comment by sabado** [comment]
Switch to Netplan and mantain a bug from '17, put automatic updates by default in server edition. Did you guys have worked on Microsoft or what?

**Comment by vorlonofportland** [comment]
On November 12, 2020 7:23:05 AM PST, Ramiro Ricoy <notifications@github.com> wrote:
>Switch to Netplan and mantain a bug from '17, put automatic updates by
>default in server edition. Did you guys have worked on Microsoft or
>what?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/CanonicalLtd/netplan/pull/34#issuecomment-726145743

You're replying to a closed issue.  Maybe you want to open a new issue for whatever bug you're referring to?
-- 
Steve Langasek

**Comment by n-cc** [comment]
d4884cfd40e1e33540b274371c3272df6595d22c adds support for networkd's `ConfigureWithoutCarrier` option via the `ignore-carrier` key.

**Pull request closed without merging**